File: habit-tracker/src/main/java/com/habit/habit_tracker/controller/HabitLogController.java
package com.habit.habit_tracker.controller;

import java.time.LocalDate;
import java.time.format.DateTimeParseException;

import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PatchMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;

import com.habit.habit_tracker.domain.DailyHabitLog;
import com.habit.habit_tracker.domain.WeeklyHabitLog;
import com.habit.habit_tracker.dto.logs.daily.DHLRequest;
import com.habit.habit_tracker.dto.logs.daily.DHLResponse;
import com.habit.habit_tracker.dto.logs.weekly.WHLRequest;
import com.habit.habit_tracker.dto.logs.weekly.WHLResponse;
import com.habit.habit_tracker.exception.ApiRequestException;
import com.habit.habit_tracker.service.DailyHabitLogService;
import com.habit.habit_tracker.service.WeeklyHabitLogService;

import jakarta.validation.Valid;

@RestController
@RequestMapping("api/v1/logs")
public class HabitLogController {

    private final DailyHabitLogService dailyHabitLogService;
    private final WeeklyHabitLogService weeklyHabitLogService;

    public HabitLogController(
            DailyHabitLogService dailyHabitLogService,
            WeeklyHabitLogService weeklyHabitLogService) {
        this.dailyHabitLogService = dailyHabitLogService;
        this.weeklyHabitLogService = weeklyHabitLogService;
    }

    @PatchMapping("/dhl")
    public ResponseEntity<DHLResponse> createOrUpdateDailyHabitLog(@Valid @RequestBody DHLRequest request) {
        DailyHabitLog dhl = dailyHabitLogService.createOrUpdateDailyHabitLog(request);
        DHLResponse response = new DHLResponse(
                dhl.getHabit().getId(),
                dhl.getMinutesDone(),
                dhl.getNotes(),
                dhl.getDate());

        return ResponseEntity.ok(response);
    }

    @GetMapping("/dhl")
    public ResponseEntity<DHLResponse> getDailyHabitLog(
            @RequestParam("habitId") long habitId,
            @RequestParam("date") String dateString) {
        try {
            LocalDate date = LocalDate.parse(dateString);
            DailyHabitLog dhl = dailyHabitLogService.getDailyHabitLog(habitId, date);

            DHLResponse response = new DHLResponse(
                    dhl.getHabit().getId(),
                    dhl.getMinutesDone(),
                    dhl.getNotes(),
                    dhl.getDate());

            return ResponseEntity.ok(response);
        } catch (DateTimeParseException e) {
            throw new ApiRequestException("Invalid date format. Please use YYYY-MM-DD.", HttpStatus.BAD_REQUEST);
        }
    }

    @PatchMapping("/week")
    public ResponseEntity<WHLResponse> createOrUpdateWeeklyHabitLog(@Valid @RequestBody WHLRequest request) {
        WeeklyHabitLog whl = weeklyHabitLogService.createOrUpdateWeeklyHabitLog(request);
        WHLResponse response = new WHLResponse(
                whl.getId(),
                whl.getHabit().getId(),
                whl.getWeeklyGoal(),
                whl.getDailyGoal(),
                whl.getMinutesDone(),
                whl.getWeeklyImbalance(),
                whl.getNotes(),
                whl.getStartDate(),
                whl.getEndDate());

        return ResponseEntity.ok(response);
    }
}

File: habit-tracker/src/main/kotlin/com/habit/habit_tracker/controller/HabitLogController.kt
package com.habit.habit_tracker.controller

import com.habit.habit_tracker.dto.logs.request.daily.DailyHabitLogRequest
import com.habit.habit_tracker.dto.logs.response.daily.DailyHabitLogResponse
import com.habit.habit_tracker.dto.logs.request.weekly.WeeklyHabitLogRequest
import com.habit.habit_tracker.dto.logs.response.weekly.WeeklyHabitLogResponse
import com.habit.habit_tracker.dto.logs.response.FullHabitLogsForWeek
import com.habit.habit_tracker.mapper.DailyHabitLogMapper
import com.habit.habit_tracker.mapper.WeeklyHabitLogMapper
import com.habit.habit_tracker.mapper.FullHabitLogsForWeekMapper
import com.habit.habit_tracker.exception.ApiRequestException
import com.habit.habit_tracker.service.DailyHabitLogService
import com.habit.habit_tracker.service.WeeklyHabitLogService
import com.habit.habit_tracker.service.HabitService
import jakarta.validation.Valid
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.*
import java.time.LocalDate
import java.time.format.DateTimeParseException

@RestController
@RequestMapping("api/v1/logs")
class HabitLogController(
    private val dailyHabitLogService: DailyHabitLogService,
    private val weeklyHabitLogService: WeeklyHabitLogService,
    private val habitService: HabitService
) {

    @PatchMapping("/daily/{habitId}")
    fun createOrUpdateDailyHabitLog(
        @PathVariable habitId: Long, 
        @Valid @RequestBody request: DailyHabitLogRequest
        ): ResponseEntity<DailyHabitLogResponse> {
        val dailyHabitLog = dailyHabitLogService.createOrUpdateDailyHabitLog(habitId, request)
        val response = DailyHabitLogMapper.toDailyHabitLogResponse(dailyHabitLog)
        return ResponseEntity.ok(response)
    }

    @PatchMapping("/weekly/{habitId}")
    fun createOrUpdateWeeklyHabitLog(
        @PathVariable habitId: Long, 
        @Valid @RequestBody request: WeeklyHabitLogRequest
        ): ResponseEntity<WeeklyHabitLogResponse> {
        val weeklyHabitLog = weeklyHabitLogService.createOrUpdateWeeklyHabitLog(habitId, request)
        val response = WeeklyHabitLogMapper.toWeeklyHabitLogResponse(weeklyHabitLog)
        return ResponseEntity.ok(response)
    }

    @GetMapping("/daily")
    fun getDailyHabitLog(
        @RequestParam("habitId") habitId: Long,
        @RequestParam("date") dateString: String
    ): ResponseEntity<DailyHabitLogResponse> {
        return try {
            val date = LocalDate.parse(dateString)
            val dailyHabitLog = dailyHabitLogService.getDailyHabitLog(habitId, date)
            val response = DailyHabitLogMapper.toDailyHabitLogResponse(dailyHabitLog)
            ResponseEntity.ok(response)
        } catch (e: DateTimeParseException) {
            throw ApiRequestException("Invalid date format. Please use YYYY-MM-DD.", HttpStatus.BAD_REQUEST)
        }
    }

    @GetMapping("/week")
    fun getFullHabitLogsForWeek(
        @RequestParam("habitId") habitId: Long,
        @RequestParam("weekStart") weekStartString: String,
        @RequestParam("weekEnd") weekEndString: String
    ): ResponseEntity<FullHabitLogsForWeek> {
        return try {
            val weekStart = LocalDate.parse(weekStartString)
            val weekEnd = LocalDate.parse(weekEndString)

            val dailyHabitLogs = dailyHabitLogService.getDailyHabitLogsForWeek(habitId, weekStart, weekEnd)
            val weeklyHabitLog = weeklyHabitLogService.getWeeklyHabitLog(habitId, weekStart, weekEnd)
            val habit = habitService.getHabit(habitId)

            val fullHabitLogsForWeek = FullHabitLogsForWeekMapper.toFullHabitLogsForWeek(
                dailyHabitLogs, 
                weeklyHabitLog, 
                habit
            )

            ResponseEntity.ok(fullHabitLogsForWeek)

        } catch (e: DateTimeParseException) {
            throw ApiRequestException("Invalid date format. Please use YYYY-MM-DD.", HttpStatus.BAD_REQUEST)
        }
    }
}
